README.md
# Hyperskill_Kotlin_SmartCalculator

This was the 8th project I completed as part of Jetbrain Academy's "Kotlin Basics" course

It is rated as a challenging difficulty project. The goal is to make a calculator that does all kinds of calculations!
It was a lot of fun. I learned some nice maths, and got to get really involved with regex

From the /help command:
```
        This program does math!
        
        Here are some things it can do:
        
    Assign a variable to a number:
        [variable_name]=[a_number]
        a = 2
        b = 3
        
    Assign a variable to another variable that's already defined:
        [variable_name]=[another_variable_name]
        a = 2 
        b = a 
        now, both a and b are equal to 2
    
    Assign a variable to an expression:
        
        c = 2*a+b^6
        this only works if 'a' and 'b' are already defined.
        
    Re-assign variables.
        Input:
        a = 1 
        b = a 
        a = 3 
        Note, b is now also equal to 3
        You can learn more about this by typing /variables
        
    Evaluate expressions using variables or numbers
        Supported operators: +, -, *, /
        1 + 2 
        output: 3
        3 * 5 
        output: 15
        
    Evaluate more complicated expressions, taking care to follow expression priority:
        2 - 1 * 8 + 6 / 2 
        output: -3
        3 + 8 * ((4 + 3) * 2 + 1) - 6 / (2 + 1) 
        output: 121
        
    That's about everything.
        
    Command List:
        /commands   - Shows all commands available in the command list.
        /help       - Prints the help message.
        /list       - Lists currently assigned variables.
        /variables  - Gives you more information about how the program handles variables
        /exit       - Turns off the program
```

I have also included the tests that Hyperskill uses to determine if a project is complete.


CHANGELOG.md
# Hyperskill_Kotlin_SmartCalculator

I decided to move the blog-post style ramblings into this CHANGELOG, on recommendation of a friend. Thanks, Loren!

### Stage 1/8: 2+2

The first stage of the course had lots of lessons listed as prerequisites in order to complete the first stage. From "Kotlin basics" like var declaration and working with strings to things like logging, memory allocation and self-documenting code.

The task of the first stage is to take an input of 2 integers on a single line, and print their sum. That's a pretty straightforward task! So, I decided to plan ahead, make the calculator singleton and add error handling.... just in case.

It's probably "too much code" for right now, but it might save me pain in the future, who knows.

### Stage 2/8: 2+2+

This next stage adds a basic control loop and other simple boxes to check like:
  -make the program continue running until '\exit' is entered
  -allow empty lines & single integer inputs

I tried to plan ahead by breaking everything into smaller tasks. I'm sure soon I'll need subtract, multiply, and divide functions.
I also made sum() allow for multiple operands (though there's still a check for only 2 operands imposed by the project)

I figure I could have written `println(operands.sumOf { it })` 
instead of ```
val sum = operands.sumOf { it }
println(sum)```
But I figure it's easier to parse this way.

. o O ( I wonder if there's a better way to add a list of numbers than `List<Int>.sumOf. {it}`?)

### Stage 3/8: Count them allocation

Well this is a good example of the benefits of thinking ahead.
Added a single-line function for a '\help' command and removed the single-line restriction on the add function.

### Stage 4/8: Add subtractions

This gave me more trouble than expecting.
Now that a valid input is no longer space separated integers, but full phrases with operators and operands, I can't resort to MutableList<Int>.sumOf {it}
I needed to make new functions to parse the input string, find the operators, and then perform the associated function.
Eliminating redundant operators was simple enough with String.replace(), but what gave me trouble was the actual function I used to reduce an input string to a result.
Originally I thought it was best to run .math() recursively.
    1. Find a valid operator
    2. Split the input string into left and right MutableList<String>s
    3. Depending on the operator, call the appropriate function on left.math() and right.math(), 
    4. .math() would then call recursively until each list of Strings became a single string.
    
But calling recursively like that doesn't actually follow the rules of... Maths.
So I went back and just used a while loop to iterate over the input list until it's reduced to a single value
I don't like how right now I'm using
```
this[operatorIndex] = result.toString()
this.removeAt(operatorIndex + 1)
this.removeAt(operatorIndex - 1)
```
But it works. More importantly, it should allow for multiplication & division to be added easily in the future.

### Stage 5/8 Error!

Repeating the theme of the last stage, this one gave me more trouble than I expected.
All I needed to do, really. was change the logic around to work  with regex on strings instead of breaking down into charSequences like I had before.
That and change the input logic to detect invalid commands/expressions.

But I wanted to plan ahead! I wanted to make sure that, when it came time to add `*` and `/`, I would be ready.
And, by trying to do to things at once, I made each of those things harder. 
I ended up writing and re-writing my code two or three times and caused lots of unnecessary pain.
That's okay though. I've learned my lesson and will only need to repeat it a few dozen times for it to really sink in ^_^

I'm not totally satisfied with this version but here are the changes
- Removed the ability to detect extra whitespace(though it can be re-added fairly easily)
- Things work with regex/Strings now instead of `MutableList<Char>`
- Added checks for invalid commands/expression
- Opened this changelog in a different editor and found/corrected lots of typos

Todo:
- There's a single line of code specifically to remove the unnecessary `+` when a user includes it in a single-operand expression, and I want so badly to remove it/fold it into another function but doing so would require me to either re-asses what an "operand" is or change something else in the logic and darn it I just want things to work for right now!

### Stage 6/8 Variables

Cool. Cool. Cool.
okay so back in Stage 3 I think? at some point the description said "you code should not throw Exceptions" 
Anyway, I totally misunderstood that as not using Exceptions at all, even for control flow purposes. But no. Actually. That's silly. I can totally do that.
Which is why things went from throwing relevant Exceptions to `println([exception text here])` which, really messed with control flow.

Also wow, Regex continues to cause me problems, but I'm getting a little more comfortable with it.

Oh yeah, the changes:
- You can now assign and sum variables (variables can only contain letters)
- Back to catching thrown exceptions for control flow oh my goodness it's so much easier this way.
- Renamed a few functions to make them clearer
- Got rid of the line of code that was bothering me when I realized `.toInt()` clears the extraneous `+`

### Stage 7/8 I've got the power

We did it!
We added multiplication! And operator preference!
And we did it in a way I kinda totally didn't expect, so the "planning ahead" bit kinda didn't work but that's OK.
So now, when we do maths, we instead convert the input to a LinkedList/Deque based on reverse polish notation RPN. Then pop items from the front of the Deque onto a stack. if the evaluated object is an operator, pop two entries off the working stack and perform the associated function for the evaluated object.

Smart Calculator/task/src/calculator/Main.kt
package calculator

import java.util.*

class CalculatorException: RuntimeException{
    constructor() : super()
    constructor(message: String?) : super(message)
}
object Calculator {
    private val priority3Regex = """\^""".toRegex()
    private val priority2Regex = """[*/]""".toRegex()
    private val priority1Regex = """[+-]+""".toRegex()
    private val operatorRegex = """($priority3Regex|$priority2Regex|$priority1Regex)""".toRegex()
    private val variableRegex = """[a-zA-Z]+""".toRegex()
    private val numberRegex = """[-+]?\d+""".toRegex()
    private val operandRegex = """($numberRegex|$variableRegex)""".toRegex()
    private val expressionRegex = """\(* *$operandRegex *\)*( *$operatorRegex *\(* *$operandRegex *\)*)*""".toRegex()
    private val assignmentRegex = """$variableRegex *= *(($operandRegex|$expressionRegex))""".toRegex()
    private val assignmentLeftRegex = """$variableRegex *=.*""".toRegex()
    private val commandRegex = """/(commands|list|exit|help|variables)| """.toRegex()
    private val variables = mutableMapOf<String, String>()

    fun run() {
        println("Enter a command, assign a variable, or enter an expression to evaluate. \nYou can type \u001B[0;32m\\commands\u001B[m for the command list, \u001B[0;32m\\help\u001B[0m for more information, or \u001B[0;32m\\exit\u001B[0m to shut down the calculator.")
        do {
            val input = readln().trim()
            try {
                when {
                    (input == "") -> {}
                    input.startsWith('/') -> {
                        if (input.matches(commandRegex)) runCommand(input)
                        else throw CalculatorException("Unknown Command")
                    }
                    input.contains("=") -> {
                        if (!assignmentLeftRegex.matches(input)) throw CalculatorException("Invalid identifier")
                        if (!assignmentRegex.matches(input)) throw CalculatorException("Invalid assignment")
                        mapIdentifier(input)
                    }
                    else -> {
                        if (input.isExpression()) println(input.mathRPN())
                        else throw CalculatorException("Invalid expression")
                    }
                }
            } catch (e: CalculatorException) {
                println(e.message)
            }
        } while (input != "/exit")
    }

    private fun runCommand(command: String) {
        when (command) {
            "/help" -> showHelp()
            "/variables" -> aboutVariables()
            "/list" -> listVariables()
            "/exit" -> exit()
            "/commands" -> commandList()
        }
    }

    private fun commandList() = println("""
    Command List:
    /commands   - Shows all commands available in the command list.
    /help       - Prints the help message.
    /list       - Lists currently assigned variables.
    /variables  - Gives you more information about how the program handles variables
    /exit       - Turns off the program
    """.trimIndent())

    private fun listVariables() {
        variables.forEach{ (key, value) -> println("$key = $value")}
    }
    private fun showHelp() = println("""
        This program does math!
        
        Here are some things it can do:
        
    Assign a variable to a number:
        [variable_name]=[a_number]
        a = 2
        b = 3
        
    Assign a variable to another variable that's already defined:
        [variable_name]=[another_variable_name]
        a = 2 
        b = a 
        now, both a and b are equal to 2
    
    Assign a variable to an expression:
        
        c = 2*a+b^6
        this only works if 'a' and 'b' are already defined.
        
    Re-assign variables.
        Input:
        a = 1 
        b = a 
        a = 3 
        Note, b is now also equal to 3
        You can learn more about this by typing /variables
        
    Evaluate expressions using variables or numbers
        Supported operators: +, -, *, /
        1 + 2 
        output: 3
        3 * 5 
        output: 15
        
    Evaluate more complicated expressions, taking care to follow expression priority:
        2 - 1 * 8 + 6 / 2 
        output: -3
        3 + 8 * ((4 + 3) * 2 + 1) - 6 / (2 + 1) 
        output: 121
        
    That's about everything.
    """.trimIndent())
    private fun aboutVariables() = println("""
        Some notes about variables:
        
        variable names can only contain letters from the latin alphabet
        a2, _myVar, and Th1s are all invalid variables
        
        variables are case sensitive, so if
        a = 1
        but you try to evaluate A + 6
        the output will be unknown variable!
        
        Variables are also evaluated "lazily"
        meaning that if A = 1 and B = A
        when A changes, so does B
        
    """.trimIndent())
    private fun exit() = println("Bye!")

    private fun mapIdentifier(assignment: String) {
        if (!assignment.matches(assignmentRegex)) throw RuntimeException("entered mapIdentifier illegally")
        val (identifierKey, identifierValue) = assignment.split("=").map { it.trim() }
        if (variableRegex.matches(identifierValue) && !variables.containsKey(identifierValue)) throw CalculatorException("Unknown variable")
        if (identifierValue.containsSelfReference(identifierKey)) throw CalculatorException("Cannot assign $identifierKey to $identifierValue because $identifierValue contains $identifierKey")
        variables[identifierKey] = identifierValue
    }
    private fun String.containsSelfReference(assignee: String): Boolean {
        return when {
            this == assignee -> true
            numberRegex.matches(this) -> false
            variableRegex.matches(this) -> variables[this]?.containsSelfReference(assignee) ?: throw CalculatorException("Variable in Expression not in variables List!")
            expressionRegex.matches(this) -> variableRegex.findAll(this).any { it.value.containsSelfReference(assignee) }
            else -> throw CalculatorException("DEBUG: Calculator Exception: $this does not match any valid Regex")
        }
    }

    private fun String.maths(operator: String = "?", second: String = ""): String {
        val a = this.decode().toBigInteger()
        val b = second.decode().toBigInteger()
        val result = when (operator.decode()) {
            "-" -> a - b
            "+" -> a + b
            "*" -> a * b
            "/" -> a / b
            "^" -> (a.toBigDecimal().pow(b.toInt())).toBigInteger()
            "?" -> throw CalculatorException("Did not pass operator to maths function")
            else -> throw CalculatorException("Something unexpected in maths function")
        }
        return result.toString()
    }
    private fun String.toExpressionList(): MutableList<String> {
        var workingString = this.replace(" ", "")
        val workingList = mutableListOf<String>()
        var searchingForOperand = true
        do {
            var match: MatchResult?
            if (searchingForOperand) {
                match = "\\(.*".toRegex().matchEntire(workingString)
                if (match != null) {
                    match = "\\(+".toRegex().find(workingString)!!
                    repeat(match.value.length) {
                        workingString = workingString.substring(1)
                        workingList.add("(")
                    }
                }

                if (!"$operandRegex.*".toRegex().matches(workingString)) throw CalculatorException("invalid exp")
                match = operandRegex.find(workingString) ?: throw CalculatorException("invalid exp")
                workingList.add(match.value)
                workingString = workingString.substring(match.value.length)

                match = "\\).*".toRegex().matchEntire(workingString)
                if (match != null) {
                    match = "\\)+".toRegex().find(workingString)!!
                    repeat(match.value.length) {
                        workingString = workingString.substring(1)
                        workingList.add(")")
                    }
                }
                searchingForOperand = false
                continue
            }
            else {
                if ("\\(.*".toRegex().matches(workingString)) throw CalculatorException("invalid exp")
                if (!"$operatorRegex.*".toRegex().matches(workingString)) throw CalculatorException("invalid exp")
                match = operatorRegex.find(workingString) ?: throw CalculatorException("invalid exp")
                workingList.add(match.value)
                workingString = workingString.substring(match.value.length)
                if ("\\).*".toRegex().matches(workingString)) throw CalculatorException("invalid exp")
                searchingForOperand = true
            }
        } while (workingString.isNotEmpty())
        return workingList
    }
    private fun String.mathRPN(): String {
        val postfix: LinkedList<String> = this.toPostfixList()
        val workingStack: Stack<String> = Stack()
        while (postfix.isNotEmpty()) {
            val inHand = postfix.pop()?: throw CalculatorException("Pop on empty stack")
            if (operandRegex.matches(inHand)) {
                workingStack.push(inHand)
                continue
            }
            if (operatorRegex.matches(inHand)) {
                val secondOperand = workingStack.pop()?: throw CalculatorException("Pop on empty stack")
                val firstOperand = workingStack.pop()?: throw CalculatorException("Pop on empty stack")
                workingStack.push(firstOperand.maths(operator = inHand,second = secondOperand))
                continue
            }
            throw CalculatorException("inHand string neither operator or operand... How'd that get in there?")
        }
        return workingStack.pop().decode()
    }
    private fun String.toPostfixList(): LinkedList<String> {
        val expressionList = this.toExpressionList()
        val tempStack: LinkedList<String> = LinkedList()
        val postfixList = LinkedList<String>()
        for (item in expressionList) {
            when {
                operandRegex.matches(item) -> postfixList.add(item)

                tempStack.isEmpty() -> tempStack.push(item)
                (tempStack.peek() == "(") -> if(item == ")") tempStack.pop() else tempStack.push(item)
                (item.priority() > tempStack.peek().priority()) -> tempStack.push(item)
                (item == "(") -> tempStack.push(item)

                (item.priority() <= tempStack.peek().priority()) -> postfixList.addAll(tempStack.popStack(item))
                (item == ")") -> postfixList.addAll(tempStack.popStack(item))
                else -> {}
            }
        }
        postfixList.addAll(tempStack.popStack())
        return postfixList
    }
    private fun LinkedList<String>.popStack(compareVal: String = ""): Queue<String> {
        val toReturn: Queue<String> = LinkedList()
        while (this.isNotEmpty()) {
            if (this.peek().priority() >= compareVal.priority()) {
                toReturn.add(this.pop())
            }
            else if (this.peek() == "(") {
                if (compareVal == ")") {
                    this.pop()
                }
                break
            }
            else break
        } //TODO this can get cleaned up a little
        if (compareVal != "" && compareVal != ")") this.push(compareVal)
        return toReturn
    }
    private fun String.priority(): Int {
        return when (true) {
            "\\(".toRegex().matches(this) -> -2
            "\\)".toRegex().matches(this) -> -1
            priority3Regex.matches(this) -> 3
            priority2Regex.matches(this) -> 2
            priority1Regex.matches(this) -> 1
            else -> 0
        }
    }
    private fun String.isExpression(): Boolean {
        if (this.count{it == '('} != this.count{it == ')'}) throw CalculatorException("Invalid # of parenthesis")
        if (!expressionRegex.matches(this)) throw CalculatorException("Invalid identifier")
        return true
    }

    private fun String.decode(): String {
        var returnString = this
        if (operandRegex.matches(this)) {
            while (!numberRegex.matches(returnString)) {
                if (!variables.containsKey(this)) {
                    throw CalculatorException("Unknown variable")
                }
                returnString = variables[returnString]?: throw CalculatorException("Uh-oh! Fun new error")
                if (expressionRegex.matches(returnString)) returnString = returnString.mathRPN()
            }
        }
        if (priority1Regex.matches(returnString)) {
            while (!"[-+]".toRegex().matches(returnString)) {
                returnString = returnString.replace("--", "+")
                    .replace("++", "+")
                    .replace("+-", "-")
                    .replace("-+", "-")
            }
        }
        return returnString
    } //Turns variables into numbers and removes extra + or -
}

fun main() {
    Calculator.run()
}
